File: clients/nodejs/src/queues.ts
/**
 * Worker task queue surface — wire types, request body builders, and
 * response parsers for the `/v1/queues` endpoints.
 *
 * External workers drive this API: they poll a named queue for leased
 * tasks, execute them, and settle each one via complete/fail (the
 * lease token from the poll response authenticates heartbeat /
 * complete / fail). Chain `worker` steps and workflow continuation
 * tasks flow through the same queues.
 *
 * The {@link ActeonClient} class carries the actual HTTP methods
 * (kept inline because TypeScript doesn't have Python-style mixins
 * and the class already aggregates every other surface); the
 * `Worker` class in `worker.ts` wraps them in a poll loop.
 */

/**
 * Lifecycle status of a worker task.
 */
export type WorkerTaskStatus =
  | "pending"
  | "leased"
  | "completed"
  | "failed"
  | "cancelled";

/**
 * A task on a worker queue.
 */
export interface WorkerTask {
  /** Unique task ID. */
  taskId: string;
  /** Queue the task is routed through. */
  queue: string;
  /** Action type for worker handler dispatch. */
  actionType: string;
  /** Task payload. */
  payload: unknown;
  /** Lifecycle status. */
  status: WorkerTaskStatus;
  /** Delivery attempt (1-based once leased). */
  attempt: number;
  /** Maximum delivery attempts. */
  maxAttempts: number;
  /** Lease token (present in poll responses; required for heartbeat /
   *  complete / fail). */
  leaseToken?: string;
  /** When the current lease expires. */
  leaseExpiresAt?: string;
  /** Result reported by the worker. */
  result?: unknown;
  /** Error reported on failure. */
  error?: string;
  /** Owning chain execution, for chain `worker` steps. */
  chainId?: string;
  /** Owning workflow execution, for workflow continuation tasks. */
  workflowExecutionId?: string;
  /** When the task was enqueued. */
  createdAt: string;
  /** When the task was last updated. */
  updatedAt: string;
}

/**
 * Parse a WorkerTask from API response.
 */
export function parseWorkerTask(data: Record<string, unknown>): WorkerTask {
  return {
    taskId: data.task_id as string,
    queue: data.queue as string,
    actionType: data.action_type as string,
    payload: data.payload,
    status: data.status as WorkerTaskStatus,
    attempt: data.attempt as number,
    maxAttempts: data.max_attempts as number,
    leaseToken: data.lease_token as string | undefined,
    leaseExpiresAt: data.lease_expires_at as string | undefined,
    result: data.result,
    error: data.error as string | undefined,
    chainId: data.chain_id as string | undefined,
    workflowExecutionId: data.workflow_execution_id as string | undefined,
    createdAt: data.created_at as string,
    updatedAt: data.updated_at as string,
  };
}

/**
 * Parse a `{ tasks: [...] }` envelope (poll + list responses).
 */
export function parseTaskListResponse(
  data: Record<string, unknown>,
): WorkerTask[] {
  const tasks = (data.tasks as Record<string, unknown>[]) ?? [];
  return tasks.map(parseWorkerTask);
}

/** Options accepted by `ActeonClient.pollTasks`. */
export interface PollTasksOptions {
  /** Maximum number of tasks to lease in one poll. */
  maxTasks?: number;
  /** Lease duration in seconds for the returned tasks. */
  leaseSeconds?: number;
  /** Stable worker identity recorded on the lease. */
  workerId?: string;
}

/** Build the `POST /v1/queues/{queue}/poll` request body. */
export function pollTasksBody(
  namespace: string,
  tenant: string,
  options: PollTasksOptions = {},
): Record<string, unknown> {
  const body: Record<string, unknown> = { namespace, tenant };
  if (options.maxTasks !== undefined) body.max_tasks = options.maxTasks;
  if (options.leaseSeconds !== undefined) {
    body.lease_seconds = options.leaseSeconds;
  }
  if (options.workerId !== undefined) body.worker_id = options.workerId;
  return body;
}

/** Options accepted by `ActeonClient.enqueueTask`. */
export interface EnqueueTaskOptions {
  /** Maximum delivery attempts before the task fails terminally. */
  maxAttempts?: number;
}

/** Build the `POST /v1/queues/{queue}/tasks` request body. */
export function enqueueTaskBody(
  namespace: string,
  tenant: string,
  actionType: string,
  payload: unknown,
  options: EnqueueTaskOptions = {},
): Record<string, unknown> {
  const body: Record<string, unknown> = {
    namespace,
    tenant,
    action_type: actionType,
    payload,
  };
  if (options.maxAttempts !== undefined) {
    body.max_attempts = options.maxAttempts;
  }
  return body;
}

/** Build the `POST /v1/queues/tasks/{task_id}/heartbeat` request body. */
export function heartbeatTaskBody(
  namespace: string,
  tenant: string,
  leaseToken: string,
  extendSeconds?: number,
): Record<string, unknown> {
  const body: Record<string, unknown> = {
    namespace,
    tenant,
    lease_token: leaseToken,
  };
  if (extendSeconds !== undefined) body.extend_seconds = extendSeconds;
  return body;
}

/** Build the `POST /v1/queues/tasks/{task_id}/complete` request body. */
export function completeTaskBody(
  namespace: string,
  tenant: string,
  leaseToken: string,
  result: unknown,
): Record<string, unknown> {
  return { namespace, tenant, lease_token: leaseToken, result };
}

/** Build the `POST /v1/queues/tasks/{task_id}/fail` request body. */
export function failTaskBody(
  namespace: string,
  tenant: string,
  leaseToken: string,
  error: string,
  retryable: boolean,
): Record<string, unknown> {
  return { namespace, tenant, lease_token: leaseToken, error, retryable };
}

File: crates/client/src/queues.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{ActeonClient, Error};

/// A worker task delivered through a named queue.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct WorkerTask {
    /// Unique task ID.
    pub task_id: String,
    /// Queue the task is routed through.
    pub queue: String,
    /// Action type for worker handler dispatch.
    pub action_type: String,
    /// Task payload.
    pub payload: Value,
    /// Lifecycle status (`"pending"`, `"leased"`, `"completed"`, `"failed"`,
    /// `"cancelled"`).
    pub status: String,
    /// Delivery attempt (1-based once leased).
    pub attempt: u32,
    /// Maximum delivery attempts.
    pub max_attempts: u32,
    /// Lease token; required for heartbeat / complete / fail.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lease_token: Option<String>,
    /// When the current lease expires.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lease_expires_at: Option<String>,
    /// Result reported by the worker.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    /// Error reported on failure.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    /// Owning chain execution for chain `worker` steps.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub chain_id: Option<String>,
    /// Owning workflow execution for workflow continuation tasks.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub workflow_execution_id: Option<String>,
    /// When the task was enqueued.
    pub created_at: String,
    /// When the task was last updated.
    pub updated_at: String,
}

/// Response from polling or listing a queue.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PollQueueResponse {
    /// Leased (or listed) tasks.
    pub tasks: Vec<WorkerTask>,
}

/// Options for polling a queue.
#[derive(Debug, Clone, Default)]
pub struct PollOptions {
    /// Maximum tasks to lease in one poll (default 1).
    pub max_tasks: Option<usize>,
    /// Lease duration in seconds (default 60, max 3600).
    pub lease_seconds: Option<u64>,
    /// Identifier of the polling worker.
    pub worker_id: Option<String>,
}

impl ActeonClient {
    /// Enqueue a task on a named worker queue.
    pub async fn enqueue_task(
        &self,
        namespace: &str,
        tenant: &str,
        queue: &str,
        action_type: &str,
        payload: Value,
        max_attempts: Option<u32>,
    ) -> Result<WorkerTask, Error> {
        let url = format!("{}/v1/queues/{queue}/tasks", self.base_url);
        let body = serde_json::json!({
            "namespace": namespace,
            "tenant": tenant,
            "action_type": action_type,
            "payload": payload,
            "max_attempts": max_attempts,
        });
        self.queue_post(&url, &body, "enqueue task").await
    }

    /// Lease pending tasks from a queue.
    pub async fn poll_tasks(
        &self,
        namespace: &str,
        tenant: &str,
        queue: &str,
        options: &PollOptions,
    ) -> Result<Vec<WorkerTask>, Error> {
        let url = format!("{}/v1/queues/{queue}/poll", self.base_url);
        let body = serde_json::json!({
            "namespace": namespace,
            "tenant": tenant,
            "max_tasks": options.max_tasks,
            "lease_seconds": options.lease_seconds,
            "worker_id": options.worker_id,
        });
        let response: PollQueueResponse = self.queue_post(&url, &body, "poll queue").await?;
        Ok(response.tasks)
    }

    /// Extend the lease on a task.
    pub async fn heartbeat_task(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
        lease_token: &str,
        extend_seconds: Option<u64>,
    ) -> Result<WorkerTask, Error> {
        let url = format!("{}/v1/queues/tasks/{task_id}/heartbeat", self.base_url);
        let body = serde_json::json!({
            "namespace": namespace,
            "tenant": tenant,
            "lease_token": lease_token,
            "extend_seconds": extend_seconds,
        });
        self.queue_post(&url, &body, "heartbeat task").await
    }

    /// Complete a leased task with a result.
    pub async fn complete_task(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
        lease_token: &str,
        result: Value,
    ) -> Result<WorkerTask, Error> {
        let url = format!("{}/v1/queues/tasks/{task_id}/complete", self.base_url);
        let body = serde_json::json!({
            "namespace": namespace,
            "tenant": tenant,
            "lease_token": lease_token,
            "result": result,
        });
        self.queue_post(&url, &body, "complete task").await
    }

    /// Fail a leased task. Retryable failures within the attempt budget are
    /// re-queued with backoff.
    pub async fn fail_task(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
        lease_token: &str,
        error: &str,
        retryable: bool,
    ) -> Result<WorkerTask, Error> {
        let url = format!("{}/v1/queues/tasks/{task_id}/fail", self.base_url);
        let body = serde_json::json!({
            "namespace": namespace,
            "tenant": tenant,
            "lease_token": lease_token,
            "error": error,
            "retryable": retryable,
        });
        self.queue_post(&url, &body, "fail task").await
    }

    /// Get a worker task by ID.
    pub async fn get_task(
        &self,
        namespace: &str,
        tenant: &str,
        task_id: &str,
    ) -> Result<WorkerTask, Error> {
        let url = format!("{}/v1/queues/tasks/{task_id}", self.base_url);
        let response = self
            .add_auth(self.client.get(&url))
            .query(&[("namespace", namespace), ("tenant", tenant)])
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;

        if response.status().is_success() {
            response
                .json::<WorkerTask>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(Error::Http {
                status: response.status().as_u16(),
                message: format!("Failed to get task: {}", response.status()),
            })
        }
    }

    /// List tasks on a queue, optionally filtered by status.
    pub async fn list_tasks(
        &self,
        namespace: &str,
        tenant: &str,
        queue: &str,
        status: Option<&str>,
    ) -> Result<Vec<WorkerTask>, Error> {
        let url = format!("{}/v1/queues/{queue}/tasks", self.base_url);
        let mut query: Vec<(&str, &str)> = vec![("namespace", namespace), ("tenant", tenant)];
        if let Some(s) = status {
            query.push(("status", s));
        }
        let response = self
            .add_auth(self.client.get(&url))
            .query(&query)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;

        if response.status().is_success() {
            let result = response
                .json::<PollQueueResponse>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))?;
            Ok(result.tasks)
        } else {
            Err(Error::Http {
                status: response.status().as_u16(),
                message: format!("Failed to list tasks: {}", response.status()),
            })
        }
    }

    /// Shared POST-and-decode helper for queue endpoints.
    async fn queue_post<T: serde::de::DeserializeOwned>(
        &self,
        url: &str,
        body: &Value,
        context: &str,
    ) -> Result<T, Error> {
        let response = self
            .add_auth(self.client.post(url))
            .json(body)
            .send()
            .await
            .map_err(|e| Error::Connection(e.to_string()))?;

        if response.status().is_success() {
            response
                .json::<T>()
                .await
                .map_err(|e| Error::Deserialization(e.to_string()))
        } else {
            Err(Error::Http {
                status: response.status().as_u16(),
                message: format!("Failed to {context}: {}", response.status()),
            })
        }
    }
}
